internal/blocksync/reactor.go

```diff
@@ -554,6 +442,158 @@ FOR_LOOP:
+	state := bcR.initialState
+
+	lastHundred := time.Now()
+	lastRate := 0.0
+
+	didProcessCh := make(chan struct{}, 1)
+
+	initialCommitHasExtensions := (bcR.initialState.LastBlockHeight > 0 && bcR.store.LoadBlockExtendedCommit(bcR.initialState.LastBlockHeight) != nil)
+
+	return didProcessCh, trySyncTicker, statusUpdateTicker, switchToConsensusTicker, blocksSynced, chainID, state, lastHundred, lastRate, initialCommitHasExtensions
+}
+
+// handlePeerError processes an error received from a peer.
+// If the peer that caused the error is still connected, it stops the peer and logs the error.
+// This function is used in the poolRoutine of the Reactor to handle errors received from peers.
+func (bcR *Reactor) handlePeerError(err peerError) {
+	peer := bcR.Switch.Peers().Get(err.peerID)
+	if peer != nil {
+		bcR.Switch.StopPeerForError(peer, err)
+	}
+}
+
+// handleBlockRequest sends a block request to a peer.
+func (bcR *Reactor) handleBlockRequest(request BlockRequest) {
+	peer := bcR.Switch.Peers().Get(request.PeerID)
+	if peer == nil {
+		return
+	}
+	queued := peer.TrySend(p2p.Envelope{
+		ChannelID: BlocksyncChannel,
+		Message:   &bcproto.BlockRequest{Height: request.Height},
+	})
+	if !queued {
+		bcR.Logger.Debug("Send queue is full, drop block request", "peer", peer.ID(), "height", request.Height)
+	}
+}
+
+// processBlocks processes two blocks and their extended commit.
+func (bcR *Reactor) processBlocks(first *types.Block, firstParts *types.PartSet, extCommit *types.ExtendedCommit, second *types.Block, firstID types.BlockID, state sm.State, blocksSynced *uint64, lastRate *float64, lastHundred *time.Time) (sm.State, error) {
+	// TODO: batch saves so we dont persist to disk every block
+	if state.ConsensusParams.Feature.VoteExtensionsEnabled(first.Height) {
+		bcR.store.SaveBlockWithExtendedCommit(first, firstParts, extCommit)
+	} else {
+		// We use LastCommit here instead of extCommit. extCommit is not
+		// guaranteed to be populated by the peer if extensions are not enabled.
+		// Currently, the peer should provide an extCommit even if the vote extension data are absent
+		// but this may change so using second.LastCommit is safer.
+		bcR.store.SaveBlock(first, firstParts, second.LastCommit)
+	}
+
+	// TODO: same thing for app - but we would need a way to
+	// get the hash without persisting the state
+	state, err := bcR.blockExec.ApplyVerifiedBlock(state, firstID, first)
+	if err != nil {
+		// TODO This is bad, are we zombie?
+		panic(fmt.Sprintf("Failed to process committed block (%d:%X): %v", first.Height, first.Hash(), err))
+	}
+	bcR.metrics.recordBlockMetrics(first)
+	*blocksSynced++
+
+	if *blocksSynced%100 == 0 {
+		_, height, maxPeerHeight := bcR.pool.IsCaughtUp()
+		*lastRate = 0.9**lastRate + 0.1*(100/time.Since(*lastHundred).Seconds())
+		bcR.Logger.Info("Block Sync Rate", "height", height, "max_peer_height", maxPeerHeight, "blocks/s", *lastRate)
+		*lastHundred = time.Now()
+	}
+
+	return state, err
```

Comment:
why are we returning an err if on line 523 we panic if `err != nil`

Reply:
I wonder if this is what I broke, I'll go through the code again.  thank you!

Reply:
@melekes I've changed this so that there's no panic.  But I am not fully certain that we DON'T want it to panic.  Could you let me know your thoughts?

Reply:
We want to panic! The panic must stay.

Reply:
🖖